tensorstore/kvstore/s3/s3_resource.h

```diff
@@ -65,6 +61,85 @@ struct S3RequesterPaysResource
   }
 };
 
+
+/// Specifies S3 endpoint url
+struct S3Endpoint : public internal::ContextResourceTraits<S3Endpoint> {
```

Comment:
Not everything in a Spec needs to be a resource. Resources are properties that come from the context and which may be shared across different instances. Thus, the cache and concurrency objects are resources.

For the s3 driver I suspect that you want the ability to specify different endpoints on a per-driver basis. So
I think that you just make this a std::string in the spec.

Reply:
OK I've done so.

Re: the concurrency + rate-limiting resources, I've re-used the GCS versions, for the moment. I'd imagine there should be S3 specific versions at some point, but I'd like to avoid factoring out common logic until major functionality is working.